### DISCUSSION TITLE
Imports & Alternative sync

### DISCUSSION THREAD

**Discussion by Salahuddeen** [davidmyersdev/octo#135] [Q&A]
Hi, Cool project. 

is there a feature pipeline for Importing from other Editors or even just importing markdown files into octo ? 

Also Can it be run without firebase?

**Comment by davidmyersdev** [comment]
Hello, Salahuddeen!

Thanks so much for taking the time to start this discussion.

There is a plan to support markdown file imports, but it has not been started just yet. Can you share any more details about what you would expect from that feature?

As for alternative backends, this is absolutely something I'd like to support. Do you have some ideas for this as well? I've previously considered adding Git as a supported backend.

**Comment by Salahuddeen** [comment]
So I currently have multiple journals, Joplin, Obsidian, Raw Markdown, OneNote, Evernote, Samsung Notes. 

Ideally, I'd like to upload a file, define the Data e.g. Pass a JSON struct. That would tell Octo how to read the files and those would then be converted to Octo's format. 

As for the Backend, Is there any reason you didn't implement Git as a backend? 
A feature of Obsidian that is RAW md files are written to the filesystem. That way if Obsidian ever goes away, I have MD files that I can then use for archiving etc or even importing into a new app etc. 

Have you ever looked at something similar to that?

**Reply by davidmyersdev** [reply]
> Ideally, I'd like to upload a file, define the Data e.g. Pass a JSON struct.

This is an interesting idea. I like it! Have an example struct you could share?

> Is there any reason you didn't implement Git as a backend?

Primarily because [octo.app](https://octo.app) is meant to serve as an all-in-one solution without _requiring_ configuration from the end user. I've always planned to add support for it though!

> if Obsidian ever goes away, I have MD files that I can then use for archiving etc

Technically, all files in Octo are stored in IndexedDB in your browser first (whether you are logged in or not), and you can always export them as JSON at the bottom of the settings page. This entire workflow could be improved though, and I don't blame you if you prefer a privately-owned backend. 😄

**Reply by Salahuddeen** [reply]
That all makes sense, I will post an example struct a little later.

Is there a feature board or timeline for feature implementations? 

I think Using local files would be really good, or even an S3 plug where MD files are written directly to an S3 bucket that the user owns. Joplin has something similar, that lets us maintain our own data store.

**Reply by davidmyersdev** [reply]
> Is there a feature board or timeline for feature implementations?

The project board is available [under the projects tab](https://github.com/writewithocto/octo/projects/1). I need to do a refresh of the cards on that board (some things are missing), and I don't have a specific timeline for items at the moment. I changed jobs recently, and coming up to speed there has been my primary focus over the last month or so. To give you an idea though, here are the features I'm currently working on in (roughly) priority order.

1. Release v1 of cross-publishing feature (post directly to dev.to, hashnode.com, and medium.com from Octo)
2. Add support for file attachments (via Google Cloud Storage)
3. Create an extension API for [Ink](https://github.com/writewithocto/ink) (the markdown component for Octo)
4. Add support for _some_ liquid tags (starting with CodeSandbox / depends on the Ink extension API)
5. Experiment with GPT-3 integration for writing assistance

There are also other little things I'm working on.

- Add a doc metadata bar (word count, read time, etc)
- Add a focus mode (hides menu bars)
- Clean up / refactor Contexts feature
- Implement Storybook to start testing UI components

> I think Using local files would be really good

Unfortunately, working with local files is difficult for a PWA. We don't have access to the filesystem or really any persistent storage ([yet](https://developer.mozilla.org/en-US/docs/Web/API/File_System_Access_API)). I like your idea of operating on a cloud storage bucket though. I will likely need to create an abstraction layer around docs for that to work. There are admittedly some implicit assumptions about the doc structure, but I might be able to work around those with some markdown metadata fields and custom pack/unpack functions. I'll also need to introduce a mechanism for choosing the provider.

All that said, there is a fair amount of work to get to this point, and it will likely take some time. Any ideas or PRs are absolutely appreciated, and this conversation has already been very helpful! 😄

**Comment by Salahuddeen** [comment]
Also is there a contribution guide? I may have a crack at adding some things to the project/ helping out.

**Reply by davidmyersdev** [reply]
No contribution guide yet, as I've primarily been the only one working on it. What kind of information are you looking for? I can likely give you an overview of the areas you are interested in and then write that up for others too.

**Reply by Salahuddeen** [reply]
I'll take a look at the source code, also been busy with work and stuff. Is there an easier way to get into contact with you for questions?

**Reply by Salahuddeen** [reply]
Also, How did the Spike on Nuxt turn out?

**Reply by davidmyersdev** [reply]
This is the best place right now. I like to keep conversations in public in case others would benefit, but if you prefer something more direct, my [Twitter DMs](https://twitter.com/voraciousdev) would work too.
